channeldb/control_tower.go

```diff
@@ -0,0 +1,487 @@
+			return err
+		}
+
+		paymentStatus := fetchPaymentStatus(bucket)
+
+		switch {
+
+		// Our records show the payment as unknown, meaning it never
+		// should have left the switch.
+		case paymentStatus == StatusUnknown:
+			updateErr = ErrPaymentNotInitiated
+			return nil
+
+		// A failed response was received for an InFlight payment, mark
+		// it as Failed to allow subsequent attempts.
+		case paymentStatus == StatusInFlight:
+
+		// The payment succeeded previously, and we are now reporting
+		// that it has failed. Leave the status as suceeded, but alert
+		// the user that something is wrong.
+		case paymentStatus == StatusSucceeded:
+			updateErr = ErrPaymentAlreadySucceeded
+			return nil
+
+		// The payment was already failed, and we are now reporting that
+		// it has failed again. Leave the status as failed, but alert
+		// the user that something is wrong.
+		case paymentStatus == StatusFailed:
+			updateErr = ErrPaymentAlreadyFailed
+			return nil
+
+		default:
+			updateErr = ErrUnknownPaymentStatus
+			return nil
+		}
+
+		// Put the failure reason in the bucket for record keeping.
+		v := []byte{byte(reason)}
+		return bucket.Put(paymentFailInfoKey, v)
+	})
+	if err != nil {
+		return err
+	}
+
+	return updateErr
+}
+
+// createPaymentBucket creates or fetches the sub-bucket assigned to this
+// payment hash.
+func createPaymentBucket(tx *bbolt.Tx, paymentHash lntypes.Hash) (
+	*bbolt.Bucket, error) {
+
+	payments, err := tx.CreateBucketIfNotExists(paymentsRootBucket)
+	if err != nil {
+		return nil, err
+	}
+
+	return payments.CreateBucketIfNotExists(paymentHash[:])
+}
+
+// fetchPaymentBucket fetches the sub-bucket assigned to this payment hash. If
+// the bucket does not exist, it returns ErrPaymentNotInitiated.
+func fetchPaymentBucket(tx *bbolt.Tx, paymentHash lntypes.Hash) (
+	*bbolt.Bucket, error) {
+
+	payments := tx.Bucket(paymentsRootBucket)
+	if payments == nil {
+		return nil, ErrPaymentNotInitiated
+	}
+
+	bucket := payments.Bucket(paymentHash[:])
+	if bucket == nil {
+		return nil, ErrPaymentNotInitiated
+	}
+
+	return bucket, nil
+
+}
+
+// nextPaymentSequence returns the next sequence number to store for a new
+// payment.
+func nextPaymentSequence(tx *bbolt.Tx) ([]byte, error) {
+	payments, err := tx.CreateBucketIfNotExists(paymentsRootBucket)
+	if err != nil {
+		return nil, err
+	}
+
+	seq, err := payments.NextSequence()
+	if err != nil {
+		return nil, err
+	}
+
+	b := make([]byte, 8)
+	binary.BigEndian.PutUint64(b, seq)
+	return b, nil
+}
+
+// fetchPaymentStatus fetches the payment status of the payment. If the payment
+// isn't found, it will default to "StatusUnknown".
+func fetchPaymentStatus(bucket *bbolt.Bucket) PaymentStatus {
+	if bucket.Get(paymentSettleInfoKey) != nil {
+		return StatusSucceeded
+	}
+
+	if bucket.Get(paymentFailInfoKey) != nil {
+		return StatusFailed
+	}
+
+	if bucket.Get(paymentCreationInfoKey) != nil {
+		return StatusInFlight
+	}
+
+	return StatusUnknown
```

Comment:
This is an error condition isn't it? We create the payment bucket in the same tx as the creation info key.

Reply:
Won't this be the default, valid case after calling `createPaymentBucket` when initiating the payment?

Reply:
It feels a bit strange to define this enum value while it really is something that should never be returned except during the creation. Maybe `createPaymentBucket` can return whether the bucket was already there. If it wasn't, `fetchPaymentStatus` and acting on the return value can be skipped.

Reply:
I don't see any big advantages of doing that, as we want to handle the `unknown` and `failed` statuses the same way.